feat: prevent importing deprecated UI components in new code

n3ps opened the pull request:
## **Description**

To increase migration into @metamask/design-system-react package, prevent new code from using the deprecated components during pre-commit.

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/37519?quickstart=1)

## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and accurately)
-->

CHANGELOG entry: null

## **Related issues**

Fixes:

## **Manual testing steps**

0. Ensure husky is installed
1. Create a new file that imports one of the deprecated UI components
2. Reference that file (make sure it's used)
3. Git commit these changes

Should see a Failed message in the console as part of the pre-commit hook

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.


<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Adds a pre-commit fitness rule to block new imports of deprecated UI components.
> 
> - **Fitness functions**:
>   - **New rule**: Add `preventDeprecatedImports` to `rules/index.ts` to block importing deprecated UI components in new files with a clear error message and docs link.
>   - **Implementation**: Introduce `rules/prevent-deprecated-imports.ts` that scans added diff lines for `from '...<deprecated-path>...'` across a defined list of deprecated `component-library/*` and `ui/*` paths; skips checks for changes under `development/fitness-functions/`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 9f049a1cbf0df2b06f19aa859df293a30710c477. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

georgewrmarshall commented on development/fitness-functions/rules/prevent-deprecated-imports.ts:
```diff
@@ -0,0 +1,43 @@
+  'component-library/popover',
+  'component-library/text',
+  'component-library/text-field-search',
+  'component-library/textarea',
```
non-blocking: should we also add deprecated components from the ui folder? e.g.
```
'ui/box',
'ui/button',
'ui/check-box',
'ui/typography',
'ui/form-field',
'ui/chip',
'ui/text-field',
'ui/identicon',
'ui/popover',
'ui/site-origin',
'ui/url-icon',
'ui/textarea',
'ui/menu',
'ui/icon-with-fallback',
'ui/icon-border',
'ui/callout',
'ui/actionable-message'
```

Got this list from searching `@deprecated` in `ui/components/ui`

<img width="363" height="283" alt="Image" src="https://github.com/user-attachments/assets/efcfe0a2-8ecb-48e7-9775-d55334b34348" />

n3ps replied:
Updated!

georgewrmarshall reviewed (commented):
This is excellent! Would it be possible to add manual testing steps so I can test this 🙏

n3ps committed:
feat: prevent importing deprecated UI components in new code

Add fitness function to prevent new imports from deprecated component-library and ui paths.
Developers should use @metamask/design-system-react instead.

Fix: Skip checking the fitness-functions directory itself to avoid false positives
when modifying the deprecated paths list.

georgewrmarshall committed:
chore: updating deprecated components list

georgewrmarshall commented on development/fitness-functions/rules/prevent-deprecated-imports.ts:
```diff
@@ -0,0 +1,66 @@
+  'component-library/modal-content/deprecated/modal-content',
+  'component-library/modal-header/deprecated/modal-header',
+  'component-library/text',
+  'component-library/text-field-search/deprecated',
```
We have a few nested deprecated components that are still JS components but weren't replaced these could likely be replaced pretty easily. I'll look into it

georgewrmarshall reviewed (approved):
Great work! I pushed a small update to the deprecated component list hope that is ok. Also tested locally using manual testing steps! 🔥 

<img width="483" height="220" alt="Image" src="https://github.com/user-attachments/assets/f7443ad6-1ea1-46a0-b433-a74e89467186" />
